src/main/java/org/codeforamerica/messaging/services/MessageService.java

```diff
@@ -108,21 +95,33 @@ public Message saveMessage(MessageRequest messageRequest, MessageBatch messageBa
 
     public void sendMessage(Long messageId) {
         log.info("Sending message #{}", messageId);
-        try {
-            Message message = messageRepository.findById(messageId).get();
-            if (message.needToSendSms()) {
-                SmsMessage sentSmsMessage = this.smsService.sendSmsMessage(message.getToPhone(), message.getSmsBody());
+        Message message = messageRepository.findById(messageId).get();
+        TemplateVariant templateVariant = message.getTemplateVariant();
+        Map<String, String> templateParams = message.getTemplateParams();
+        if (message.needToSendSms()) {
+            try {
+                String smsBody = templateVariant.build(TemplateVariant::getSmsBody, templateParams);
+                SmsMessage sentSmsMessage = this.smsService.sendSmsMessage(message.getToPhone(), smsBody);
                 message.setSmsMessage(sentSmsMessage);
                 messageRepository.save(message);
+            } catch (IOException e) {
+                log.error("Error processing SMS template: " + templateVariant.getId());
+            } catch (Exception e) {
+                log.error("Error sending SMS job", e);
             }
-            if (message.needToSendEmail()) {
-                EmailMessage sentEmailMessage = this.emailService.sendEmailMessage(message.getToEmail(), message.getEmailBody(), message.getSubject());
+        }
+        if (message.needToSendEmail()) {
+            try {
+                String subject = templateVariant.build(TemplateVariant::getSubject, templateParams);
+                String emailBody = templateVariant.build(TemplateVariant::getEmailBody, templateParams);
+                EmailMessage sentEmailMessage = this.emailService.sendEmailMessage(message.getToEmail(), emailBody, subject);
                 message.setEmailMessage(sentEmailMessage);
                 messageRepository.save(message);
+            } catch (IOException e) {
+                log.error("Error processing email templates: " + templateVariant.getId());
```

Comment:
Minor, this may no longer be relevant. Plural here vs singular in the SMS case.

Reply:
In terms of minor distinctions, in my mind there are technically two email templates, the body and the subject.